### PR TITLE
Deny undocumented unsafe blocks

### PR DESCRIPTION
**Pull request by liuw**
Tighten the requirement for documented unsafe blocks by using Clippy's `undocumented_unsafe_blocks` lint.

The slight downside of this lint is it is a bit rigid in the form of the safety comment, but the upside is it is very thorough and basically leaves no stone unturned.

Fix some existing comments to meet the style requirement, and more importantly, document all unsafe blocks which did not have safety comments before.

### CODE REVIEW

File: event_monitor/src/lib.rs [hunk 1]
```diff
@@ -16,8 +16,10 @@ static mut MONITOR: Option<(File, Instant)> = None;
 /// This function must only be called once from the main process before any threads
 /// are created to avoid race conditions
 pub fn set_monitor(file: File) -> Result<(), std::io::Error> {
+    // SAFETY: access global variable
```
Comment (likebreath) [inline event_monitor/src/lib.rs:19]: It needs to be further clarified, as accessing a global mutable variable is why the code is unsafe. 

I think what makes it safe here is we manually "ensured" the gloabal variable (`MONITOR`) will be written only once and all reads will happen after this variable is written.

File: event_monitor/src/lib.rs [hunk 2]
```diff
@@ -26,6 +28,7 @@ pub fn set_monitor(file: File) -> Result<(), std::io::Error> {
     if ret < 0 {
         return Err(std::io::Error::last_os_error());
     }
+    // SAFETY: access global variable
```
Comment (likebreath) [inline event_monitor/src/lib.rs:31]: Ditto

File: event_monitor/src/lib.rs [hunk 3]
```diff
@@ -41,6 +44,7 @@ struct Event<'a> {
 }
 
 pub fn event_log(source: &str, event: &str, properties: Option<&HashMap<Cow<str>, Cow<str>>>) {
+    // SAFETY: access global variable
```
Comment (likebreath) [inline event_monitor/src/lib.rs:47]: Ditto

File: tracer/src/tracer.rs [hunk 4]
```diff
@@ -101,11 +102,14 @@ struct TraceEvent {
 
 pub fn trace_point_log(event: &'static str) {
     let trace_event = TraceEvent {
+        // SAFETY: access global variable
```
Comment (likebreath) [inline tracer/src/tracer.rs:105]: I think the safety comments around `TRACER` need to be further clarified. @rbradford

**Review by likebreath — COMMENTED** [review]
Looking good. My main comments are around how to document safety around using mutable global variables.

File: vmm/src/vm.rs [hunk 5]
```diff
@@ -2932,10 +2932,12 @@ impl GuestDebuggable for Vm {
 
         #[cfg(feature = "tdx")]
         {
-            if self.config.lock().unwrap().tdx.is_some() {
-                return Err(GuestDebuggableError::Coredump(anyhow!(
-                    "Coredump not possible with TDX VM"
-                )));
+            if let Some(ref platform) = self.config.lock().unwrap().platform {
```
[the remark this answers is not in the document]
Comment (liuw) [reply vmm/src/vm.rs:2936]: That's right.

This is caught by `cargo clippy --workspace --bins --examples --benches --all-features --all-targets -- -D clippy::undocumented_unsafe_blocks`

File: event_monitor/src/lib.rs (hunk 1 above)
Comment (liuw) [reply event_monitor/src/lib.rs]: I took a stab at this. Please review.

File: tracer/src/tracer.rs (hunk 4 above)
Comment (liuw) [reply tracer/src/tracer.rs]: I updated this patch. I tried to reason about this module but admittedly I didn't look too deeply.

File: tracer/src/tracer.rs [hunk 6]
```diff
@@ -134,16 +140,17 @@ impl TraceBlock {
 
 impl Drop for TraceBlock {
     fn drop(&mut self) {
+        // SAFETY: start is a variable that's local to the current thread
+        let start = unsafe { TRACER.get().unwrap().start };
```
Comment (rbradford) [inline tracer/src/tracer.rs:144]: `// SAFETY: start has been initialised as part of initialising the value of TRACER`

File: tracer/src/tracer.rs [hunk 7]
```diff
@@ -134,16 +140,17 @@ impl TraceBlock {
...
             event: self.event,
-            end_timestamp: Some(
-                Instant::now().duration_since(unsafe { TRACER.get().unwrap().start }),
-            ),
+            end_timestamp: Some(Instant::now().duration_since(start)),
+            // SAFETY: thread_depth() returns a number local to the current thread
             depth: unsafe { TRACER.get().unwrap().thread_depth() },
         };
+        // SAFETY: add_event and decrease_thread_depth access storage local to
+        // the current thread only
```
Comment (rbradford) [inline tracer/src/tracer.rs:153]: This makes it sound like `thread_local!` so perhaps:

`// SAFETY: add_event and decrease_thread_depth access current thread only specific data`

File: tracer/src/tracer.rs [hunk 8]
```diff
@@ -122,6 +126,8 @@ pub struct TraceBlock {
 
 impl TraceBlock {
     pub fn new(event: &'static str) -> Self {
+        // SAFETY: increase_thread_depth accesses storage local to
+        // the current thread only
```
Comment (rbradford) [inline tracer/src/tracer.rs:130]: `// SAFETY: increase_thread_depth accesses current thread only specific data`

File: tracer/src/tracer.rs [hunk 9]
```diff
@@ -105,11 +106,14 @@ struct TraceEvent {
 
 pub fn trace_point_log(event: &'static str) {
     let trace_event = TraceEvent {
+        // SAFETY: start is a variable that's local to the current thread
```
Comment (rbradford) [inline tracer/src/tracer.rs:109]: `// SAFETY: start has been initialised as part of initialising the value of TRACER`

File: tracer/src/tracer.rs [hunk 10]
```diff
@@ -105,11 +106,14 @@ struct TraceEvent {
...
         timestamp: Instant::now().duration_since(unsafe { TRACER.get().unwrap().start }),
         event,
         end_timestamp: None,
+        // SAFETY: thread_depth() returns a number local to the current thread
```
Comment (rbradford) [inline tracer/src/tracer.rs:113]: `// SAFETY: thread_depth accesses current thread only specific data`

File: tracer/src/tracer.rs [hunk 11]
```diff
@@ -105,11 +106,14 @@ struct TraceEvent {
...
         depth: unsafe { TRACER.get().unwrap().thread_depth() },
     };
+    // SAFETY: add_event adds an event to a storage that's only accessible to the current thread
```
Comment (rbradford) [inline tracer/src/tracer.rs:116]: `// SAFETY: add_event accesses current thread only specific data`

**Review by rbradford — APPROVED** [review]

**Pull request merged**